[quant][core] Add Embedding and EmbeddingBag reference module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73436

Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

jerryzh168 committed:
[quant][core] Add Embedding and EmbeddingBag reference module

Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
Will be tested later, in the final PR that deprecates the current convert implementation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
Will be tested later, in the final PR that deprecates the current convert implementation

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

vkuzo reviewed (commented):
should we have a correctness test for the new modules?

jerryzh168 commented:
yes, planning to do that in a later PR that changes the current convert.py to something new, it will be a large PR so trying to land these things first to reduce the size. we may fix errors in that PR as well

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
Will be tested later, in the final PR that deprecates the current convert implementation

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
added some tests, please take a look again, thanks!

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

vkuzo commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -1591,3 +1597,66 @@ def test_rnn(self):
+        num_embeddings = 10
+        embedding_dim = 3
+        # embedding input
+        ex = torch.LongTensor([[1,2,4,5],[4,3,2,9]])
```
nit: lint

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][core] Add Embedding and EmbeddingBag reference module"


Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).
